crt/debugmalloc.c

```diff
@@ -0,0 +1,538 @@
+
+#define RETURN_ADDRESS(nr) \
+  __builtin_extract_return_addr(__builtin_return_address (nr))
+
+extern bool __crt_enable_debug_malloc;
+
+/* weak musl libc functions used by this module */
+extern void* libc_malloc(size_t size);
+extern void libc_free(void* ptr);
+extern void* libc_memalign(size_t alignment, size_t size);
+extern void* libc_calloc(size_t nmemb, size_t size);
+extern void* libc_realloc(void* ptr, size_t size);
+extern size_t libc_malloc_usable_size(void* ptr);
+
+MYST_NOINLINE
+size_t myst_backtrace_impl(void** start_frame, void** buffer, size_t size)
+{
+    void** frame = start_frame;
+    size_t n = 0;
+    const int prot = PROT_READ;
+
+    while (n < size)
+    {
+        if (myst_mcheck(frame, 1, PROT_READ) != 0 ||
+            myst_mcheck(frame[1], 1, PROT_READ) != 0)
+        {
+            break;
+        }
+
+        buffer[n++] = frame[1];
+        frame = (void**)*frame;
+    }
+
+    return n;
+}
+
+size_t myst_backtrace(void** buffer, size_t size)
+{
+    return myst_backtrace_impl(__builtin_frame_address(0), buffer, size);
+}
+
+static void _dump_backtrace(void* addrs[], int num_addrs)
+{
+    size_t i;
+
+    for (i = 0; i < num_addrs; i++)
+    {
+        /* ATTN: address symbol names not known */
+        fprintf(stderr, "%p\n", addrs[i]);
+    }
+}
+
+/*
+**==============================================================================
+**
+** Debug allocator:
+**
+**     This allocator checks for the following memory errors.
+**
+**         (1) Leaked blocks on program exit.
+**         (2) Memory overwrites just before/after the block.
+**         (3) Assuming blocks are zero filled (fills new blocks with 0xAA).
+**         (3) Use of free memory (fills freed blocks with 0xDD).
+**
+**     This allocator keeps in-use blocks on a linked list. Each block has the
+**     following layout.
+**
+**         [padding] [header] [user-data] [footer]
+**
+**     The padding is applied by memalign() when the alignment is non-zero.
+**
+**==============================================================================
+*/
+
+/*
+**==============================================================================
+**
+** Local definitions:
+**
+**==============================================================================
+*/
+
+#define HEADER_MAGIC1 0x185f0447c6f5440f
+#define HEADER_MAGIC2 0x56cfbed5df804061
+#define FOOTER_MAGIC 0x8bb6dcd8f4724bc7
+
+typedef struct header header_t;
+
+struct header
+{
+    /* Contains HEADER_MAGIC1 */
+    uint64_t magic1;
+
+    /* Headers are kept on a doubly-linked list */
+    header_t* next;
+    header_t* prev;
+
+    /* The alignment passed to memalign() or zero */
+    uint64_t alignment;
+
+    /* Size of user memory */
+    size_t size;
+
+    /* Return addresses obtained by _backtrace() */
+    void* addrs[BACKTRACE_MAX];
+    uint64_t num_addrs;
+
+    uint64_t padding;
+
+    /* Contains HEADER_MAGIC2 */
+    uint64_t magic2;
+
+    /* User data */
+    uint8_t data[];
```

Comment:
Does this field count towards the size of the structure?

Reply:
No, its size is zero.